das/store.go

```diff
@@ -0,0 +1,112 @@
+package das
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/namespace"
+	"golang.org/x/net/context"
+)
+
+var (
+	storePrefix   = datastore.NewKey("das")
+	checkpointKey = datastore.NewKey("checkpoint")
+)
+
+// The checkpoint store stores/loads the DASer's checkpoint to/from
+// disk using the checkpointKey. The checkpoint is stored as a struct
+// representation of the latest successfully DASed state.
+type store struct {
+	datastore.Datastore
+}
+
+// backgroundStore periodically saves current sampling state in case of DASer force quit before
+// being able to store state on exit
+type backgroundStore struct {
+	store    *store
+	getStats func(ctx context.Context) (checkpoint, error)
```

Comment:
Why is this abstracted when it is always the same function? If the issue is that it needs a handle to the coordinator, I don't understand why we can't implement the background store routine directly as a method of the coordinator (another question here: Why is the store in the coordinator, but the background store in the DASer?)

This would mean:
- There wouldn't be a need for a backgroundStore struct (assuming we also get rid of `done` here)
- `getCheckpoint` is called directly in the implementation of the `backgroundStore` routine, reducing abstraction and improving readability

Reply:
I really don't want to put `backgroundStore` into `coordinator` because its functionality  is not related to `corrdinator's` purpose. `backgroundStore` is a separate routine and needs to  communicate state over `getstats` callback so there is no race conditions. Even tho it is same pkg and it is hard to follow dependancy graph  both `coordinator` and `backgroundStore` depend on `store` and not on each other. Thats why it should be controlled by DASer, not coordinator.

Reply:
Another option is to make a backreference. `backgroundStore` may keep the reference to the `coordinator` parent and call getStats directly from it. IMO, both ways are fine

Reply:
pass funcs: store checkpointStore, getStats func(ctx context.Context) (checkpoint, error) to backgroundStore `run()` func